### PR TITLE
Porting Dynamic_Update_Slice operator from TFLite

### PR DESCRIPTION
**Pull request by rameshkunasi**
Added Dynamic_Update_Slice operator changes.

BUG=fixes https://github.com/tensorflow/tflite-micro/issues/3217

### CODE REVIEW

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/op_macros.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/micro_log.h"
+#include "tensorflow/lite/micro/micro_utils.h"
+
+namespace tflite {
+
+constexpr int kMaxDimensions = 6;
```
Comment (ddavis-2015) [inline tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: Please change the `6` to `RuntimeShape::kMaxSmallSize`.  This is so we don't have to go hunting for 6's the next time LiteRT decides to change RuntimeShape.

Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: Thank you for the review. Updated!!

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  // check output dimensions (relocated) against original dimensions
+  TF_LITE_MICRO_EXPECT_EQ(output_dims->size,
+                          tensors[kOutputTensorIndex].dims->size);
+  for (int i = 0; i < output_dims->size; i++) {
+    TF_LITE_MICRO_EXPECT_EQ(output_dims->data[i],
+                            tensors[kOutputTensorIndex].dims->data[i]);
+  }
+}
+
+// TODO(rameshkunasi): Add quantized test for dynamic update slice.
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: Are you planning to add this in the separate PR?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+constexpr int kNumInputs = 3;
+constexpr int kNumOutputs = 1;
+constexpr int kInputTensorIndex_0 = 0;
+constexpr int kInputTensorIndex_1 = 1;
+constexpr int kInputTensorIndex_2 = 2;
+constexpr int kOutputTensorIndex = 3;
+
+// min/max are used to compute scale, zero-point is 0
+template <size_t kInputSize>
+struct TestDynamicUpdateSliceParams {
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: How do you use this?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+#include "tensorflow/lite/micro/kernels/kernel_runner.h"
+#include "tensorflow/lite/micro/micro_utils.h"
+#include "tensorflow/lite/micro/test_helpers.h"
+#include "tensorflow/lite/micro/testing/micro_test.h"
+
+namespace tflite {
+namespace testing {
+namespace {
+
+// constexpr float kTestTolerance = 7.41e-03;
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: What is this for?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+TF_LITE_MICRO_TEST(DynamicUpdateSliceOpTestSimpleFloat) {
+  int32_t kInputDims_0[] = {2, 3, 3};
+  int32_t kInputDims_1[] = {2, 2, 1};
+  int32_t kInputDims_2[] = {1, 2};
+  int32_t* kInputDims[tflite::testing::kNumInputs] = {
+      kInputDims_0, kInputDims_1, kInputDims_2};
+  int kOutputDims[] = {2, 3, 3};
+
+  constexpr float kInput_0[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc:99]: Maybe I missed something here; The dimension for this is defined {2, 3, 3} above but the array here only has 9 elements? Where is the remaining half?

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  TF_LITE_ENSURE_EQ(context, SizeOfDimension(start_indices, 0),
+                    NumDimensions(operand));
+
+  TF_LITE_ENSURE_EQ(context, NumDimensions(update), NumDimensions(operand));
+  // Check that update dimensions are not larger than operand dimensions
+  for (int i = 0; i < NumDimensions(operand); ++i) {
+    TF_LITE_ENSURE(context,
+                   SizeOfDimension(update, i) <= SizeOfDimension(operand, i));
+  }
+  output->type = operand->type;
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: I'm curious what's the purpose of this? Output tensor will be deallocated right away?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  constexpr float kExpect[] = {1, 2, 3, 4, -1, 6, 7, -2, 9};
+  constexpr int kOutputCount = std::extent<decltype(kExpect)>::value;
+  float output_data[kOutputCount];
+
+  tflite::testing::TestDynamicUpdateSliceFloat(kInputDims, kInput_0, kInput_1,
+                                               kInput_2, kExpect, kOutputDims,
+                                               output_data);
+}
+
+TF_LITE_MICRO_TEST(DynamicUpdateSliceOpTestBoundaryTest) {
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc:191]: What's the purpose of this test? How is this different from  `DynamicUpdateSliceOpTestSimpleFloat`?

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+TfLiteStatus UpdateSliceRecursive(int current_dim, int max_dims,
+                                  const int32_t* output_strides,
+                                  const int32_t* update_strides,
+                                  const int32_t* update_dims_data,
+                                  const T* update_tensor_data,
+                                  const int32_t* clamped_start_indices,
+                                  T* output_tensor_data) {
+  if (current_dim == max_dims) {
+    return kTfLiteOk;
+  }
```
Comment (veblush) [inline tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: I guess this is unreachable?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 5 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc:99]: @veblush `kOutputDims` is a stand-in array for TfLiteIntArray.  The first element of the array is the number of array elements.  This occurs throughout the unit tests (using `IntArrayFromInts`).  TfLiteFloatArray is even hairier as the first element of the float array has to be rewritten as an `int` by `FloatArrayFromFloats`...

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc (hunk 6 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: @rameshkunasi I am also curious as to why the output tensor type is not already set by the TfLiteConverter?

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 2 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: Thank you for finding this. For this operator quantized tests are not required. Added remaining tests

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 3 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: Not useful. Removed

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 4 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc]: Not required. Removed

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc (hunk 6 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: Thank you for pointing this. Not required, removed this

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 7 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc:191]: It is same as **DynamicUpdateSliceOpTestSimpleFloat** except the indices. Here in this test indices used are corner case

File: tensorflow/lite/micro/kernels/dynamic_update_slice.cc (hunk 8 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice.cc]: It will, because it will be called by increasing current_dim count in recursively

File: tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc (hunk 5 above)
Comment (rameshkunasi) [reply tensorflow/lite/micro/kernels/dynamic_update_slice_test.cc:99]: Thank you for providing details @ddavis-2015
